lib/compiler/src/beam_ssa_alias.erl

```diff
@@ -0,0 +1,997 @@
+%%%
+%%% * As the beam_ssa_opt fixpoint framework doesn't provide a way for
+%%%   an optimization to be informed that the fixpoint calculation
+%%%   didn't converge within the interation limit and it is unsafe to
+%%%   do optimizations on incomplete unique/aliased information it is
+%%%   much simpler to explicitly handle it locally instead of trying
+%%%   to detect incomplete information in a hypothetical
+%%%   ssa_opt_alias_finish pass.
+%%%
+aa_fixpoint(Order, AAS) ->
+    ?DP("**** Starting fixpoint iteration ****~n"),
+    aa_fixpoint(Order, Order, AAS#aas.alias_map, AAS, ?MAX_REPETITIONS).
+
+aa_fixpoint([F|Fs], Order, OldAliasMap, AAS0=#aas{st_map=StMap}, Limit) ->
+    #b_local{name=#b_literal{val=_N},arity=_A} = F,
+    AAS1 = AAS0#aas{caller=F},
+    ?DP("-= ~p/~p =-~n", [_N, _A]),
+    {OptSt,AAS2} = aa_fun(F, map_get(F, StMap), AAS1),
+    AAS = AAS2#aas{st_map=StMap#{F => OptSt}},
+    aa_fixpoint(Fs, Order, OldAliasMap, AAS, Limit);
+aa_fixpoint([], _Order, OldAliasMap,
+            #aas{alias_map=OldAliasMap,func_db=FuncDb,st_map=StMap}, _) ->
+    ?DP("**** End of iteration ****~n"),
+    {StMap, FuncDb};
+aa_fixpoint([], _, _, #aas{func_db=FuncDb,orig_st_map=StMap}, 0) ->
+    ?DP("**** End of iteration, too many iterations ****~n"),
+    {StMap, FuncDb};
+aa_fixpoint([], Order, _OldAliasMap, AAS=#aas{alias_map=AliasMap}, Limit) ->
+    ?DP("**** Things have changed, starting next iteration ****~n"),
+    aa_fixpoint(Order, Order, AliasMap, AAS, Limit - 1).
+
+aa_fun(F, #opt_st{ssa=Linear0,anno=_Anno,args=Args}=St,
+       AAS0=#aas{alias_map=AliasMap0}) ->
+    %% Initially assume all formal parameters are unique for a
+    %% non-exported function, if we have call argument info in the
+    %% AAS, we use it. For an exported function, all arguments are
+    %% assumed to be aliased.
+    ArgsStatus = aa_get_call_args_status(Args, F, AAS0),
+    SS0 = foldl(fun({Var, Status}, Acc) ->
+                        aa_new_ssa_var(Var, Status, Acc)
+                end, #{}, ArgsStatus),
+    ?DP("@@ Args: ~p~n", [ArgsStatus]),
+    {Linear1,SS} = aa_blocks(Linear0, SS0, AAS0),
+    ?DP("SS:~n~s~n~n", [aa_format(SS)]),
+    AAS = aa_merge_call_args_status(SS, AAS0),
+
+    AliasMap = AliasMap0#{ F => SS },
+    {St#opt_st{ssa=Linear1}, AAS#aas{alias_map=AliasMap}}.
+
+%% Main entry point for the alias analysis
+aa_blocks([{L,#b_blk{is=Is0,last=T0}=Blk}|Bs0], SS0, AAS) ->
+    {Is,SS1} = aa_is(Is0, SS0, [], AAS),
+    {T,SS2} = aa_terminator(T0, SS1),
+    {Bs,SS} = aa_blocks(Bs0, SS2, AAS),
+    {[{L,Blk#b_blk{is=Is,last=T}}|Bs],SS};
+aa_blocks([], SS, _) ->
+    {[],SS}.
+
+aa_is([I=#b_set{dst=Dst,op=Op,args=Args,anno=Anno0}|Is], SS0, Acc, AAS) ->
+    SS1 = aa_new_ssa_var(Dst, unique, SS0),
+    SS = case Op of
+             {bif,Bif} ->
+                 aa_bif(Dst, Bif, Args, SS1, AAS);
+             {float,_} ->
+                 SS1;
+             {succeeded,_} ->
+                 SS1;
```

Comment:
Nitpick: I think this `case` is a bit noisy, do you think it makes sense to split the ones needing special attention and these `SS1` ones into separate groups?

Reply:
Do you mean something like:

```Erlang
`case` Op of
  %% Instructions changing the alias status
  {bif,Bif}  ->
     aa_bif(Dst, Bif, Args, SS1, AAS);
  ...
  % Instructions which don't change the alias status
  {succeeded,_} ->
    SS1;
  ...
end
```
Or something more radical? The current order comes from when I expanded `beam_ssa:op()`, but I see the beenfit of grouping them differently. What I don't want to do is to just do:

```Erlang
 ...
  % Instructions which don't change the alias status
  _ -> SS1
end
```
as I prefer a hard crash in the compiler when a new instruction is added and it's not added here.

Reply:
I was thinking the former, grouping the instructions that don't change the status together at the bottom of the list.

Reply:
done